### PR TITLE
[CARBONDATA-748] use binary search improve filter on dimension has  no inverted index

### PR DESCRIPTION
**Pull request by mayunSaicmotor**
use binary search to improve performance according to filter values' order

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1054/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1055/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1056/

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java [hunk 1]
```diff
@@ -419,6 +419,94 @@ public static int getFirstIndexUsingBinarySearch(FixedLengthDimensionDataChunk d
     return -(low + 1);
   }
 
+  public static int[] getRangeIndexUsingBinarySearch(
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: Please provide comments this method

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java [hunk 2]
```diff
@@ -150,12 +138,15 @@ private BitSet setFilterdIndexToBitSet(DimensionColumnDataChunk dimensionColumnD
...
     if (dimensionColumnDataChunk instanceof FixedLengthDimensionDataChunk) {
       byte[][] filterValues = dimColumnExecuterInfo.getFilterKeys();
-      for (int k = 0; k < filterValues.length; k++) {
-        for (int j = 0; j < numerOfRows; j++) {
-          if (dimensionColumnDataChunk.compareTo(j, filterValues[k]) == 0) {
-            bitSet.set(j);
-          }
+      for (int i = 0; i < numerOfRows; i++) {
+
+        int index = CarbonUtil.binarySearch(filterValues, 0, filterValues.length,
```
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java]: if `filterValues` size is one then we better avoid this binary search , just compare would be enough.

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java (hunk 1 above)
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: There is not much difference between `getFirstIndexUsingBinarySearch` and this method,  I remembered in your last PR you have done binary search even for getting the ranges, what happened to it, did you get any functional or performance issues?

Comment (mayunSaicmotor) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: you are right, I  really done binary search even for getting the ranges previously, but yesterday I done performance test and found the  performance is not better than current logic. the binary search range has advantage only under the condition of  data array size is very long  and the repeated data is too much. But usually the data array size is 12000 for a chunk, not too long. So  the binary search range has no advantage and I decide to keep the current logic

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java (hunk 2 above)
Comment (mayunSaicmotor) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java]: does the below is OK?

  private BitSet setFilterdIndexToBitSet(DimensionColumnDataChunk dimensionColumnDataChunk,
      int numerOfRows) {
    BitSet bitSet = new BitSet(numerOfRows);
    if (dimensionColumnDataChunk instanceof FixedLengthDimensionDataChunk) {
      byte[][] filterValues = dimColumnExecuterInfo.getFilterKeys();
      for (int i = 0; i < numerOfRows; i++) {

        if (filterValues.length > 1) {
          int index = CarbonUtil.binarySearch(filterValues, 0, filterValues.length - 1,
              dimensionColumnDataChunk.getChunkData(i));

          if (index >= 0) {
            bitSet.set(i);
          }
        } else if (filterValues.length == 1) {
          if (dimensionColumnDataChunk.compareTo(i, filterValues[0]) == 0) {
            bitSet.set(i);
          }
        } else {
          break;
        }

      }
    }
    return bitSet;
  }

**Comment by chenliang613** [comment]
@mayunSaicmotor  please change "[Carbondata 748] " to "[CARBONDATA-748]" for PR's title.

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java (hunk 2 above)
Comment (ravipesala) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java]: looks fine

**Comment by CarbonDataQA** [comment]
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1085/

File: core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java (hunk 1 above)
Comment (mayunSaicmotor) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonUtil.java]: comments  was added.  Is there anything else need to change?

**Comment by ravipesala** [comment]
LGTM

File: core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java (hunk 2 above)
Comment (mayunSaicmotor) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/executer/IncludeFilterExecuterImpl.java]: @ravipesala, If put the if clause out of the for clause, it is better?

`  private BitSet setFilterdIndexToBitSet(DimensionColumnDataChunk dimensionColumnDataChunk,
      int numerOfRows) {
    BitSet bitSet = new BitSet(numerOfRows);
    if (dimensionColumnDataChunk instanceof FixedLengthDimensionDataChunk) {
      byte[][] filterValues = dimColumnExecuterInfo.getFilterKeys();

      if (filterValues.length > 1) {
        for (int i = 0; i < numerOfRows; i++) {
          int index = CarbonUtil.binarySearch(filterValues, 0, filterValues.length - 1,
              dimensionColumnDataChunk.getChunkData(i));

          if (index >= 0) {
            bitSet.set(i);
          }
        }
      } else if (filterValues.length == 1) {
        for (int i = 0; i < numerOfRows; i++) {
          if (dimensionColumnDataChunk.compareTo(i, filterValues[0]) == 0) {
            bitSet.set(i);
          }
        }
      }
    }
    return bitSet;
  }`

**Pull request closed without merging**